tests/phpunit/tests/option/option.php

```diff
@@ -312,4 +312,151 @@ public function data_option_autoloading() {
+
+		$this->assertFalse( $updated, 'update_option should not return true when values are loosely equal.' );
+		$this->assertSame( $num_queries, get_num_queries(), 'The number of database queries should not change.' );
+	}
+
+	/**
+	 * @ticket 22192
+	 *
+	 * @covers ::update_option
+	 *
+	 * @dataProvider data_update_option_type_juggling
+	 *
+	 * @param mixed $old_value One of the values to compare.
+	 * @param mixed $new_value The other value to compare.
+	 */
+	public function test_update_option_should_hit_cache_when_loosely_equal_to_existing_value_and_cached_values_are_pulled_from_the_database( $old_value, $new_value ) {
+		add_option( 'foo', $old_value );
+		wp_cache_delete( 'alloptions', 'options' );
+		wp_load_alloptions();
+
+		$num_queries = get_num_queries();
+
+		$updated = update_option( 'foo', $new_value );
+
+		$this->assertFalse( $updated, 'update_option should not return true when values are loosely equal.' );
+		$this->assertSame( $num_queries, get_num_queries(), 'The number of database queries should not change.' );
+	}
+
+	/**
+	 * Data provider.
+	 *
+	 * @return array
+	 */
+	public function data_update_option_type_juggling() {
+		return array(
+			// Truthy.
+			array( '1', '1' ),
+			array( '1', 1 ),
+			array( '1', 1.0 ),
+			array( '1', true ),
+			array( 1, '1' ),
+			array( 1, 1 ),
+			array( 1, 1.0 ),
+			array( 1, true ),
+			array( 1.0, '1' ),
+			array( 1.0, 1 ),
+			array( 1.0, 1.0 ),
+			array( 1.0, true ),
+			array( true, '1' ),
+			array( true, 1 ),
+			array( true, 1 ),
+			array( true, true ),
+
+			// Falsey.
+			array( '0', '0' ),
+			array( '0', 0 ),
+			array( '0', 0.0 ),
+			array( '0', false ),
+			array( 0, '0' ),
+			array( 0, 0 ),
+			array( 0, 0.0 ),
+			array( 0, false ),
+			array( 0.0, '0' ),
+			array( 0.0, 0 ),
+			array( 0.0, 0.0 ),
+			array( 0.0, false ),
+		);
+	}
+
+	/**
+	 * @ticket 22192
+	 *
+	 * @covers ::update_option
+	 *
+	 * @dataProvider data_update_option_type_falsey_values
+	 *
+	 * @param mixed $old_value One of the values to compare.
+	 * @param mixed $new_value The other value to compare.
+	 * @param int   $expected  The expected result.
+	 */
+	public function test_update_option_with_falsey_values( $old_value, $new_value, $expected ) {
+		add_option( 'foo', $old_value );
+		$num_queries = get_num_queries();
+
+		$updated = update_option( 'foo', $new_value );
+
+		$this->assertSame( $expected, $updated, 'update_option should not match expeted value when values are loosely equal.' );
+		$this->assertSame( 1, get_num_queries() - $num_queries, 'The number of database queries should not change.' );
+	}
+
+	/**
+	 * Data provider.
+	 *
+	 * @return array
+	 */
+	public function data_update_option_type_falsey_values() {
+		return array(
+			array( false, '0', true ),
+			array( false, 0, true ),
+			array( false, 0.0, true ),
+			array( false, '', false ),
+			array( false, null, false ),
+			array( false, array(), true ),
+			array( false, false, false ),
+		);
+	}
```

Comment:
@felixarntz @joemcgill In relation to the BC break that was detected previously, let's discuss.

Currently, a side-PR that tests some feedback here is showing a failure in `Tests_Multisite_GetSpaceAllowed::test_get_space_allowed()` with the following dataset:
```php
array(
    'site_option'    => 0,
    'network_option' => 666,
    'expected'       => 0,
)
```
The test method's contents:
```php
update_option( 'blog_upload_space', $site_option );
update_site_option( 'blog_upload_space', $network_option );

$this->assertSame( $expected, get_space_allowed() );
```
With the changes in this PR:
- `$site_option` is `(int) 0`. This is loosely equal to `(bool) false`.
  - `update_option( 'blog_upload_space', $site_option )` will fail.
- `$network_option` is `(int) 666`.
  - `update_option( 'blog_upload_space', $network_option )` will succeed.

The `get_space_allowed()` function:
```php
function get_space_allowed() {
	$space_allowed = get_option( 'blog_upload_space' ); // Previously 0. Now false.

	if ( ! is_numeric( $space_allowed ) ) { // Previously failed. Now passes.
		$space_allowed = get_site_option( 'blog_upload_space' ); // 666.
	}

	if ( ! is_numeric( $space_allowed ) ) { // Previously failed. Still fails.
		$space_allowed = 100;
	}

	// $space_allowed was previously 0 -> Uploads disabled. It's now 666 -> BC Break.

	// Applies filters and returns. Not including here as it's not relevant to the test failure.
}
```

This goes back to @joemcgill's earlier comment that there are valid reasons for switching types.

I believe we may need to re-evaluate our interpretation of what this ticket is asking for, and then determine whether it still applies, or is achievable without breaking BC.

Thoughts?

Reply:
@costdev Hmm, can you clarify why `update_option( 'blog_upload_space', $site_option )` is failing? Shouldn't it succeed, given that the `_is_equal_database_value()` function should only be called if there already _is_ a value in the database (see https://github.com/WordPress/wordpress-develop/pull/5139/files#diff-ca21e48c46c9dd91f3e8b8837f147ae3d99457233f4f8a042ef458205d4d81e3R786)?

I may be missing something here. But to me this seems that it's not about a valid reason for switching types, but rather about that we shouldn't compare the value if the "old value" isn't an actual database value (i.e. the option is technically being _added_).

Reply:
@felixarntz Looking into this now - Will get back shortly.

Reply:
@felixarntz Can you take a look at [the side-PR](https://github.com/mukeshpanchal27/wordpress-develop/pull/25/files) here and see what you think we might have wrong?

Reply:
@costdev I took a look at this with @mukeshpanchal27 earlier today, and with this PR checked out, I'm not seeing `Tests_Multisite_GetSpaceAllowed::test_get_space_allowed()` failing currently when I run the multisite tests. Can you confirm and/or tell me what steps I need to take to reproduce that test failure?

Reply:
@costdev I reviewed https://github.com/mukeshpanchal27/wordpress-develop/pull/25#pullrequestreview-1636047365 and the code looks right - some of the tests are wrong though. Not sure where you see the `Tests_Multisite_GetSpaceAllowed::test_get_space_allowed()` failure, I don't see it on the PR.

Reply:
@felixarntz The `Tests_Multisite_GetSpaceAllowed::test_get_space_allowed()` failure was on [an earlier test run](https://github.com/mukeshpanchal27/wordpress-develop/actions/runs/6224670505/job/16893411781) which was happened to be fixed in two commits Mukesh made after I posted the above comment and before your review. 😅